### PR TITLE
Add support for left and right joins

### LINKED ISSUE #21: Support for left joins
Right now it seems like ndjson-join always does an inner join, and will delete features that are not shared between the two files. But it's very common for an external dataset to be missing data for some features (counties, states, etc.) that you still want to display in your final map.

Could ndjson-join take a parameter to support some outer joins? For a left join where a feature was missing in the "right" file, I'd imagine that d[1] could be and empty string or simply non-existent.

Apologies if this is already possible!

### PR DESCRIPTION
**Pull request by jstcki**
Fixes #21.

I also added a new test file. The results are the following:

```sh
# Left
./ndjson-join 'd.name' <(csv2json -n test/a.csv) <(csv2json -n test/c.csv) --left
[{"name":"Fred","color":"red"},{"name":"Fred","number":"42"}]
[{"name":"Alice","color":"green"},null]
[{"name":"Bob","color":"blue"},null]

# Right
./ndjson-join 'd.name' <(csv2json -n test/c.csv) <(csv2json -n test/a.csv) --right
[{"name":"Fred","number":"42"},{"name":"Fred","color":"red"}]
[null,{"name":"Alice","color":"green"}]
[null,{"name":"Bob","color":"blue"}]

# Inner (default)
./ndjson-join 'd.name' <(csv2json -n test/a.csv) <(csv2json -n test/c.csv)
[{"name":"Fred","color":"red"},{"name":"Fred","number":"42"}]
```

### CODE REVIEW

File: ndjson-join [hunk 1]
```diff
@@ -13,6 +13,8 @@ commander
     .usage("[options] [expression₀ [expression₁]] <file₀> <file₁>")
     .description("Join two newline-delimited JSON streams into a stream of pairs.")
     .option("-r, --require <name=module>", "require a module", requires, {d: undefined, i: -1})
+    .option("--left", "Perform a left join", false)
+    .option("--right", "Perform a right join", false)
```
Comment (mbostock) [inline ndjson-join]: Lowercase “perform”.

File: ndjson-join [hunk 2]
```diff
@@ -25,6 +27,8 @@ if (commander.args.length < 4) {
   commander.args.splice(1, 0, commander.args[0]);
 }
 
+var join = commander.left ? leftJoin : commander.right ? rightJoin : innerJoin;
+
```
Comment (mbostock) [inline ndjson-join]: So if both `--left` and `--right` are set, `--left` takes priority? Perhaps instead, if both are set we should do an outerJoin?

File: ndjson-join [hunk 3]
```diff
@@ -85,3 +89,31 @@ function join() {
     });
   });
 }
+
+function leftJoin() {
+  map.forEach(function(value, key) {
+    value[0].forEach(function(v0) {
+      if (value[1].length === 0) {
+        output([v0, null])
```
Comment (mbostock) [inline ndjson-join]: Missing semicolon.

File: ndjson-join [hunk 4]
```diff
@@ -85,3 +89,31 @@ function join() {
     });
   });
 }
+
+function leftJoin() {
+  map.forEach(function(value, key) {
+    value[0].forEach(function(v0) {
+      if (value[1].length === 0) {
```
Comment (mbostock) [inline ndjson-join]: I’d invert the logic like so:

```js
if (value[1].length) {
  …
} else {
  output([v0, null]);
}
```

File: ndjson-join [hunk 5]
```diff
@@ -85,3 +89,31 @@ function join() {
...
+      }
+    });
+  });
+}
+
+function rightJoin() {
+  map.forEach(function(value, key) {
+    value[1].forEach(function(v1) {
+      if (value[0].length === 0) {
+        output([null, v1])
```
Comment (mbostock) [inline ndjson-join]: Semicolon.

File: ndjson-join [hunk 6]
```diff
@@ -85,3 +89,31 @@ function join() {
...
+        });
+      }
+    });
+  });
+}
+
+function rightJoin() {
+  map.forEach(function(value, key) {
+    value[1].forEach(function(v1) {
+      if (value[0].length === 0) {
```
Comment (mbostock) [inline ndjson-join]: Invert logic as above.

**Comment by mbostock** [comment]
Good stuff! Biggest suggestion would be to support outer join either as `--left --right` or as `--outer`. (If the latter, then the arguments should be exclusive and specifying multiple types of join should exit with an error.)

Can you update the README to document the new arguments?

**Comment by jstcki** [comment]
Done! Thanks for the feedback.

One thing I wondered: wouldn't it be better to output an empty object instead of `null` for missing values? That might be easier to handle later in the chain because for example `Object.assign(null, {foo: 1})` will fail.

**Comment by mbostock** [comment]
Released in 0.3.1! Thank you!

**Pull request merged**